src/bit_encoding/decode.rs

```diff
@@ -734,25 +730,48 @@ mod tests {
         // wit_diff = comp (comp (pair wit1 wit2) jet_subtract_32) (drop iden) :: 1 -> 2^32
         // diff_is_one = comp (pair wit_diff jet_one_32) jet_eq_32             :: 1 -> 2
         // main = comp diff_is_one jet_verify                                  :: 1 -> 1
+
+        #[rustfmt::skip]
+        let bad_diff1s = vec![
+            // Above program, but with both witness nodes shared (note they have
+            // the same type and CMR)
+            vec![
+                0xda, 0xe2, 0x39, 0xa3, 0x10, 0x42, 0x0e, 0x05,
+                0x71, 0x88, 0xa3, 0x6d, 0xc4, 0x11, 0x80, 0x80
+            ],
+            // Same program but with each `witness` replaced by `comp iden witness`, which
+            // is semantically the same but buries the offending witness nodes a bit to
+            // trip up naive sharing logic.
+            vec![
+                0xde, 0x87, 0x04, 0x08, 0xe6, 0x8c, 0x41, 0x08,
+                0x38, 0x15, 0xc6, 0x22, 0x8d, 0xb7, 0x10, 0x46,
+                0x02, 0x00,
+            ],
+        ];
+        for bad_diff1 in bad_diff1s {
+            assert_program_not_deserializable::<Core>(&bad_diff1, &crate::Error::SharingNotMaximal);
+        }
+
         #[rustfmt::skip]
         let diff1s = vec![
             (
+                // Sharing corrected
                 vec![
-                    0xda, 0xe2, 0x39, 0xa3, 0x10, 0x42, 0x0e, 0x05,
-                    0x71, 0x88, 0xa3, 0x6d, 0xc4, 0x11, 0x80, 0x80
+                    0xdc, 0xee, 0x28, 0xe6, 0x8c, 0x41, 0x08, 0x38,
+                    0x15, 0xc6, 0x22, 0x8d, 0xb7, 0x10, 0x46, 0x02,
+                    0x00,
                 ],
                 // CMR not checked against C code, since C won't give us any data without witnesses
                 "a2ad9852818c0dc9307b476464cb9366c5c97896ba128f2f526b51910218293c",
                 None,
                 None,
             ),
-            // Same program but with each `witness` replaced by `comp iden witness`, which
-            // is semantically the same but will trip up naive witness-unsharing logic.
+            // Same program but with each `witness` replaced by `comp iden witness`.
             (
                 vec![
-                    0xde, 0x87, 0x04, 0x08, 0xe6, 0x8c, 0x41, 0x08,
-                    0x38, 0x15, 0xc6, 0x22, 0x8d, 0xb7, 0x10, 0x46,
-                    0x02, 0x00,
+                    0xe0, 0x28, 0x70, 0x43, 0x83, 0x00, 0xab, 0x9a,
+                    0x31, 0x04, 0x20, 0xe0, 0x57, 0x18, 0x8a, 0x36,
+                    0xdc, 0x41, 0x18, 0x08, 
                 ],
```

Comment:
750abaaa8038045683030c40c0f84da2259ae605: Does this mean we used to share witness / disconnect nodes in the encoding if they had the same type + CMR?

Reply:
Yep. Well, we used to allow it, anyway -- the old `CommitNode` would do no sharing checks, only `RedeemNode` was.

Then as it happened, our encoding would actually do it, which is why all our test vectors (which I made with the old code) had sharing that we would now consider illegitimate.

Reply:
Once we have the human-readable encoding and you can verify these binary blobs, I hope you'll feel better about these test vectors :).